### DISCUSSION TITLE
Medusa currently has "Products". Can it also have "Services"?

### DISCUSSION THREAD

**Discussion by SaadBazaz** [medusajs/medusa#2768] [Feature Requests]
(Based on the issue #2520)

## Is your feature request related to a problem? Please describe.
A project I'm building; they sell products (e.g. car rims) and also services (e.g. car wash).

## Why?
Medusa seems like it'll dominate the upcoming open-source Product space for a while. What if it could also cover "Services", and become an all-in-one E-Commerce (Products + Services) solution? We could have just one Medusa doing it all; Products, Services, User Management, CMS (Strapi), and more.

## Describe the solution you'd like
A Service could be a special kind of "Product" which can possibly have:

- No quantity (or infinite quantity)
- Special dates/times (calendar) on which it's available (Like a [booking/appointment](https://github.com/calcom/cal.com))
- Inputs from the user
- ??
Open for discussion.
This could be an official Medusa Plugin, and could have a special place in the Admin Panel.

## Describe alternatives you've considered
- Having separate forms for each service
- Using a 3rd party appointment scheduler
- Google Forms

**Comment by srindom** [comment]
Would be great to understand in what areas the current product model falls short of being able to support your needs. In the description you give above, I think there would be a reasonably straightforward path to supporting your needs. Namely:

- When adding a "service product" to cart specify the service related properties in metadata. E.g.:
  - item.metadata.date = [appointment date]
  - item.metadata.user_input = { ... object with user details }
- Quantity for the service would always be 1

This would be done with something like:
```
POST /store/carts/:id/line-items
{
  variant_id: "id to the variant representing the service",
  quantity: 1,
  metadata: {
    ...
  }
}
```

I understand that defining the properties that could be customized is not supported well right now, but would be good to understand what you think of the approach above?

**Reply by SaadBazaz** [reply]
Hmm. Well of course that's how it seems like it'll work under the hood. I guess it's more about the user experience. E.g
- being able to filter services Vs products,
- assign/connect Users to services (e.g. an appointment with the employee Paul), 
- a "Calendar" view in the medusa dashboard which gives a view of current appointments
- Subscription functionality (reoccurring payments)

This way it might work for a SaaS product, too.

**Reply by pmltechpile** [reply]
Your solution is ultimately what I have planned to implement for a client of mine that offers cooking classes. Although doable, it seems a bit hacky to what Medusa already offers and reading @SaadBazaz's idea and layout I'd love to see that be a feature as well.

**Reply by purvanova-213** [reply]
@pmltechpile were you able to implement it using medusa

**Reply by alanmpitts** [reply]
Is this still an issue for Medusa users. I have a client who operates a salon. She has asked me for a solution that covers booking makeup artists. So a calendar view. Search for availability based on date/time and artist. Offer an alternate artist if their first choice isn't available.

**Reply by PabloJVelez** [reply]
@alanmpitts It's not an issue but also not something that comes out of the box with Medusa. Your salon use case is definitely something that would require custom development.

**Reply by alanmpitts** [reply]
Thanks for the feedback @PabloJVelez

**Reply by radoslaw-sz** [reply]
hi @alanmpitts  - did you figure it out using Medusa? I am exploring options to fill this niche with some open-source framework.

**Comment by laurent512** [comment]
Would be interested about that too !

**Comment by ddresch** [comment]
In our case, it is not only the two different types of products. If you think about event business like coaches, hairdressers etc. all businesses who sell time slots (time-based services) also the above example with car wash is like that. Imagine you need to book time slots of like 1 hour etc. 

Is there already a kind of integration for use case like that?

**Comment by JonathanRiche** [comment]
Would also be interested in this as well!

**Comment by jwoa** [comment]
Same here! Been looking for something like this and it seems like we may have to go the DIY route for now.

**Comment by alightinastorm** [comment]
The configuration of services is highly domain-specific (some might need dates, some might need much more information or different kind of information)

For medusa this doesn't really make sense to implement, since there is no general purpose solution?
Services *are products*. If you need to specify additional data around the product, then use the metadata field? 

We are doing it that way at least. And our configurator (currently in development) is a mix of physical products + services (installation/construction mainly).

So far the internal tests have been successful. It requires actually developing on your own though, and medusa gives you the options to do so

**Reply by ddresch** [reply]
The interesting part is how you define a service? Is it the act of installation you sell without handling the actual date of installation, or is it possible to book a specific date for the service you offer? If you sell the installation service with for example a flat fee and nothing else, then it is a simple product add-on of course. Can you be a bit more specific how you handle it via the metadata? I'm working with booking tools for a couple of years now, and I can tell it is not only some basic fields, so I'm curious how you handle it. Thanks for your feedback!

**Reply by alightinastorm** [reply]
For calendar integration, consider using apps like cal.com

**Reply by alightinastorm** [reply]
Well, the amount of physical products depends on the customers needs (we developed a planning tool for that, similar to how IKEA does wardrobe planning).

The installation bit is currently under development, will likely integrate a calendar solution like cal.com to get bookings integrated.

**Reply by alightinastorm** [reply]
and tbh, for your mentioned use cases in your other comment, like hairdressers, etc. something like cal.com is probably sufficient, since you can also handle payments over there. Fitting medusa into that niche seems odd to me.

**Reply by ddresch** [reply]
Thanks for your quick response. You are right for a simple one-person hairdresser business, it is not worth the overhead of medusa.   The thing is my real use case is the management and selling of escape rooms which is depending on a fixed schedule and multiple (n) rooms per location. I was checking cal.com in the past already and figured it is not the perfect fit, but maybe I need another deep look and tests. 

Maybe you know the WooCommerce ecosystem, there we have a Plugin called [Bookings](https://woo.com/fr-fr/products/woocommerce-bookings/) which allows you to define so-called bookable products. As I understand cal.com I'm able to define time slots, but I would need to map the actual product with these bookings. Doable of course!

**Reply by alightinastorm** [reply]
On Cal, you can have multiple calendars, each bookable.
If you wanted to go Medusa <-> Cal, you can create a calendar per room per location.
Rooms in Locations would be Products on the Medusa end, having a calendar on the cal end

For a successful order, users *have* to have a booking selected on the calendar (your frontend should catch this at least)

**Comment by alightinastorm** [comment]
Booking, availabilities etc are a huge pain to build, so i would hope that medusa instead focuses on the core - having a general purpose solid ecommerce base!

**Comment by allendoss** [comment]
@SaadBazaz Zenbooker is a great example! I'd love to see something like this in Medusa

Test it out here: https://zenbooker.com/mobile-auto-detailing-software. Click on Book now and use zip code 10001 to test.

**Comment by rubn-g** [comment]
It would be good to, at least, have some guide or directions in documentation about how to extend medusa to have availability management for services, event tickets, etc

**Comment by NicolasGorga** [comment]
Hey guys, anyone has tried implementing services selling. I am thinking of building an ecommerce of trades, so for example electricians, plumbers, house keepers, etc can offer their services. I am struggling a bit in visualizing how i should extend the Core Commerce Modules like Product to link with Services and also Cart and Order Module.

Would love to chat about this, my email is nicogorga11@gmail.com

**Comment by jeethjoseph** [comment]
I'm wondering about the best way to extend inventory. We will need to add a date wise inventory to check availability. And update the same based on booking. Seems Hacky?

**Reply by srinivasulu-narayanam** [reply]
I am also trying to implement same are you able to move forward with this usecase?

**Reply by NicolasGorga** [reply]
Interested too

**Reply by johanguse** [reply]
Hey, I was wondering if you all have any new ideas for implementing these partners?

**Comment by radoslaw-sz** [comment]
hi all - my company starts exploring options to fill the niche with booking/rental system on top of Medusa. Something like Mercur for marketplaces, we would like to try something for booking system.
Is there still some market needs for such thing? I would love to hear also about the features and so on. You can reply here or drop the message labs@rsoftcon.com

**Comment by radoslaw-sz** [comment]
My company started working on something like this:


https://github.com/user-attachments/assets/33101e94-55ee-43f9-8b4e-7734b17374d9

**Reply by radoslaw-sz** [reply]
new showcase:

https://github.com/user-attachments/assets/728f0439-3499-448d-8930-c2c689cb1654

**Comment by radoslaw-sz** [comment]
For people who are instereted in booking functionality (rentals, event tickets, table reservations and others) - we launched our first version of booking system here: https://github.com/RSC-Labs/medusa-booking-system
